tuner: Introduce dedicated net mode

StephanDollberg opened the pull request:
Introduces the new autodetected "dedicated" mode to the net tuner.

It uses one dedicated interrupt core (vcpu) per 16 cores (vcpus). This is the
setup that we have found to deliver the best performance even over MQ. We are
already using this in the cloud on m7gd.

Additionally when the config indicates that network tuning is on and we find
ourselves in this mode `rpk redpanda start` will automatically pass a cpuset to
redpanda such that it doesn't run on the interrupt cores. Like this you get the
fully automatic and tuned experience out of the box.

Getting this automatic behaviour right isn't always possible if the user doesn't
run on a system which fully belongs to RP. Hence, we'll need to do run some
checks first to see whether it's fine to do so. In this PR these checks aren't
implemented yet. Instead, the new mode is off by default.

This allows us to already perf test and keep implementing other features that
depend on it in a piecemeal fashion. Once, we have the checks we can enable it
by default. The flag will also make rollout easier in the cloud where we
currently have a weird mixed setup and we will likely not flip it until 26.1.

Commits should be reviewable independently.

I am still open for alternative names to "dedicated".





## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

StephanDollberg committed:
tuner: Introduce dedicated tuner mode

Introduces the `dedicated` tuner mode which will use one dedicated
interrupt core per 16 vcpus. Which as per our testing gives the best
performance over Mq and the bugged Sq modes.

For now only used by the net tuner.

StephanDollberg committed:
hwloc: Add MaskToListFormat

Adds `MaskToListFormat` to convert from a cpuset hex mask to list format
as understood by seastar.

StephanDollberg committed:
tuners: Make nics.GetDefaultMode public

Such that we can use it in other places to calculate the effective
default mode.

StephanDollberg committed:
tuner: Support dedicated net tuner mode in rpk:start

This adds the second half of the dedicated net tuner mode where `rpk
start` will automatically start redpanda with the right cpuset such that
it's not running on the interrupt cores.

This is an early commit which supports the basic functionality. We will
later need to add a lot more checks to not enable this when not tuned
correctly.

StephanDollberg committed:
tuner: Make cores per interrupt core configurable

Makes the core count that we need for each interrupt core configurable
via `rpk.cores_per_dedicated_interrupt_core`.

This will for example allow us to enable dedicated mode in cloud even
for smaller tiers where we already run with a core less and thereby
reducing the k8s tax.

StephanDollberg committed:
tuner: Add flag to enable/disable dedicated mode

Disable the automatic dedicated mode by default.

This will allow us to merge the new mode even while we don't have all
the checks in place yet that are needed for a full default roll out.

Further it will simplify the rollout of this in our cloud where we run
the tuner but also do some manual changes/hacks after that. With the
extra flag less coordination is required.

We can change the default of this flag once the checks are in place but
likely only after 25.3 has branched to get some more data/experience as
we are already close to code completition.

StephanDollberg committed:
tests: Add net tuner test for dedicated mode

Adds a test for the dedicated tuner mode.

Refactors the existing test such that we can use it for Mq and dedicated
mode.

StephanDollberg committed:
tests: Some more python type fixes

StephanDollberg commented on src/go/rpk/pkg/config/weak.go:
```diff
@@ -477,6 +479,16 @@ func (rpkc *RpkNodeConfig) UnmarshalYAML(n *yaml.Node) error {
 	rpkc.Tuners.BallastFilePath = string(internal.BallastFilePath)
 	rpkc.Tuners.BallastFileSize = string(internal.BallastFileSize)
 	rpkc.Tuners.WellKnownIo = string(internal.WellKnownIo)
+	if internal.CoresPerDedicatedInterruptCore != nil {
```
@r-vasquez here I have a question from the rpk side. I couldn't find any place where we handle default values for the config properties which is why I implemented it like this.

I am not sure there is a better place or way?

StephanDollberg replied:
I see now that the way it's implemented this means that some of these values get printed into the config when doing something like enabling prod mode.

Idealy I'd just want to use these defaults internally and not have them leak out when not needed.

r-vasquez replied:
We have:

- https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/config/redpanda_yaml.go#L240 <- defaults, that pretty much are loaded if you don't have anything on disk.
- https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/config/config.go#L262 <- prod mode

What do you mean by wanting the defaults internally? are they going to be modified by users at any point? or it will always be 16?

StephanDollberg replied:
> What do you mean by wanting the defaults internally? are they going to be modified by users at any point? 

They can be modified by the user yes. But what I want to avoid is that assuming the user hasn't changed a property is that when we write out the config (because for example rpk generates a default redpanda.yaml) I don't want it to end up in the config file.

So the states are basically:
 - unset
 - explicitly set to true
 - explicitly set to false

but if it's unset I don't want the default to end up in any config written by rpk.

StephanDollberg replied:
I have pushed a change here. It's a mix of what we discussed. Like this we don't get defaults written out and usage is at least somewhat defined.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--47a4734f-17ce-407f-b92d-889fdcc9d6bf-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72993">build#72993</a></summary>


| test_class                          | test_method                                                     | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                       |
|:------------------------------------|:----------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterRateQuotaTest                | test_client_group_consume_rate_throttle_mechanism               | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#01998298-3120-4cd7-af4e-e6c41cdfe932 | FLAKY         | 13/21    | upstream reliability is '81.28119800332779'. current run reliability is '61.904761904761905'. drift is 19.37644 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism               |
| ClusterRateQuotaTest                | test_client_group_produce_rate_throttle_mechanism               | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#01998298-3122-4faf-bbb3-ec099da30f36 | FLAKY         | 13/21    | upstream reliability is '82.60427263479146'. current run reliability is '61.904761904761905'. drift is 20.69951 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_produce_rate_throttle_mechanism               |
| ClusterRateQuotaTest                | test_client_response_and_produce_throttle_mechanism             | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#0199829f-539a-442b-be5d-4a30abee00cc | FLAKY         | 15/21    | upstream reliability is '79.30501930501931'. current run reliability is '71.42857142857143'. drift is 7.87645 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism             |
| ClusterRateQuotaTest                | test_client_response_throttle_mechanism                         | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#01998298-3128-4603-b7e8-dde3a050eaa3 | FLAKY         | 14/21    | upstream reliability is '81.59695817490494'. current run reliability is '66.66666666666666'. drift is 14.93029 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest                | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#0199829f-539c-400e-8c89-aae7122d59fb | FLAKY         | 17/21    | upstream reliability is '82.35294117647058'. current run reliability is '80.95238095238095'. drift is 1.40056 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| RandomNodeOperationsTest            | test_node_operations                                            | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72993#01998298-3128-4603-b7e8-dde3a050eaa3 | FLAKY         | 20/21    | upstream reliability is '97.31343283582089'. current run reliability is '95.23809523809523'. drift is 2.07534 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                                        |
| RpkConfigTest                       | test_config_change_mode_prod                                    | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#0199829f-539a-442b-be5d-4a30abee00cc | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RpkConfigTest&test_method=test_config_change_mode_prod                                           |
| RpkConfigTest                       | test_config_set_json                                            | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#01998298-312a-4cfb-8e28-ad21ffe34531 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RpkConfigTest&test_method=test_config_set_json                                                   |
| RpkConfigTest                       | test_config_set_json                                            | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72993#0199829f-539c-400e-8c89-aae7122d59fb | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RpkConfigTest&test_method=test_config_set_json                                                   |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                                  | {"use_transactions": false}                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72993#01998298-3128-4603-b7e8-dde3a050eaa3 | FLAKY         | 16/21    | upstream reliability is '91.90140845070422'. current run reliability is '76.19047619047619'. drift is 15.71093 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                                   |
</details>
<!--/body-->

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/start.go:
```diff
@@ -452,6 +455,45 @@ func prestart(
+		// TODO: Need to add various checks here:
+		//  - Run the checkers to confirm we are tuned correctly
+		//  - Check no existing --cpuset is configured
+		//  - Check if --smp is passed and adapt accordingly (we have some leeway here)
```
I don't think that's necessary as we only need to parse the args which are passed rpk via config/flags and those are already parsed and don't really have the `-c`, `--smp` etc. duality?

StephanDollberg replied:
Well there is various ways to pass smp. There is a separate RP node config flag, then there is additional start args (then there is probably others). The latter is already getting parsed as there are already checks implemented such that you don't pass both etc. 

Maybe -c isn't handled today. I'd need to double check.

StephanDollberg replied:
It parses everything in first instance. These are the flags rpk knows about: https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/cli/redpanda/start.go#L60-L80

I.e.: for those it will check for duplicates. Everything that's not in there will be passed verbatim (I think - that code is kinda confusing).

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/start.go:
```diff
@@ -471,10 +513,25 @@ func buildRedpandaFlags(
+	// interrupt cores.
+	if y.Rpk.Tuners.TuneNetwork {
+		mask, err := checkNetDedicatedMode(fs, y)
+		if err != nil {
```
> What kind of errors happen here?

Various errors from commands that we invoke etc. 

> it doens't mean the tuner was actually run or that anything is tunable or anything like that, right? In those cases I would still want RP to start, not error out.

Yep exactly. This is what we need the additional checks for and not all outcomes will be critical (e.g. like you say when tuning didn't happen at all).

StephanDollberg replied:
No I am referring to the to be implemented checks whether it's safe to enable dedicated mode.

StephanDollberg commented on src/go/rpk/pkg/cli/redpanda/start.go:
```diff
@@ -452,6 +455,45 @@ func prestart(
+		if err != nil {
+			return "", err
+		}
+		fmt.Println("Dedicated mode detected, starting Redpanda with cpuset", mask)
```
Yep, most of the stuff is already logged from the underlying functions but I'll add more once we get to adding the checks.

StephanDollberg commented on src/go/rpk/pkg/tuners/irq/cpu_masks.go:
```diff
@@ -83,6 +84,21 @@ func (masks *cpuMasks) CPUMaskForComputations(
 	} else if mode == Mq {
 		// all available cores
 		computationsMask = cpuMask
+	} else if mode == Dedicated {
```
Yeah it's called like that right now

StephanDollberg commented on src/go/rpk/pkg/tuners/irq/cpu_masks.go:
```diff
@@ -83,6 +84,21 @@ func (masks *cpuMasks) CPUMaskForComputations(
 		// all available cores
 		computationsMask = cpuMask
+	} else if mode == Dedicated {
+		numOfCores, err := masks.GetNumberOfPUs(cpuMask)
```
Yeah this should be numOfPUs which is the term the tuner/hwloc uses.

I am not entirely decided on whether in config for example we'd want to use "cores" or "vcpus".

StephanDollberg commented on src/go/rpk/pkg/tuners/irq/cpu_masks.go:
```diff
@@ -83,6 +85,21 @@ func (masks *cpuMasks) CPUMaskForComputations(
+		if err != nil {
+			return "", err
+		}
+		rpCores := numOfCores - uint(math.Ceil(float64(numOfCores)/float64(t.CoresPerDedicatedInterruptCore)))
```
We shouldn't really end up in that condition here as we should never be in `dedicated` mode then. The only way that could happen is if someone configured `CoresPerDedicatedInterruptCore=1`. I'll will add a check for that in `GetDefaultMode`.

> Should we instead switch to "mq" mode or something?

I think we should just error out as per above. Note the only place where we can switch mode is in `GetDefaultMode`. We can't just "switch mode" in any of the other functions as things will become inconsistent then.

StephanDollberg commented on src/go/rpk/pkg/tuners/irq/irq_tuner_mode.go:
```diff
@@ -33,18 +33,23 @@ mq - distribute NIC's IRQs among all CPUs instead of binding
-  - If number of physical CPU cores per Rx HW queue is greater than 4 - use the 'sq-split' mode.
-  - Otherwise, if number of hyperthreads per Rx HW queue is greater than 4 - use the 'sq' mode.
-  - Otherwise use the 'mq' mode.
+  - Use dedicated mode if it's allowed by config and safe to do so
```
- "allowed by config": `allow_dedicated_interrupt_mode` is set to true
- "safe to do": All the future to be implemented checks pass

StephanDollberg replied:
Yeah sorry, it's in a later commit. I kinda wrote this comment with a bit of foreshadowing.

StephanDollberg commented on src/go/rpk/pkg/tuners/network/nics.go:
```diff
@@ -35,14 +35,20 @@ func getDefaultMode(
+		return mode, nil
 	}
 
 	if nic.IsBondIface() {
```
> How about this case? Too rare to care about?

No, I expect it's kinda frequent in selfhosted but it's at the bottom of the stack of things to do.

StephanDollberg committed:
tuner: Test fixups

StephanDollberg committed:
tuner: Remove some unneeded logging from GetDefaultMode

We want this function to be as light as possible and not rely on things
that might be available in k8s for example.

This logging is redundant here anyway.

StephanDollberg commented:
I have pushed a change & rebase now:
 - A few fixes as per Travis' feedback
 - Cleaned up the config usage a bit
 - Added another commit that removes some unneeded logging

As discussed in standup I'll add a follow up commit (after some other follow ups) that will also enable the new mode in rpk:start when a tuner file is passed as otherwise it's unlikely this will ever work in k8s.

r-vasquez reviewed (dismissed):
rpk configuration handling LGTM, I'll let the final approval to Travis as he delved more into it. Thanks both!

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73388-->
<!--9f817bb4-f0f9-41f4-962b-d035d60f11e6-->
<!--title-->
## Retry command for [Build#73388](https://buildkite.com/redpanda/redpanda/builds/73388)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_retention_test.py::OffsetRetentionTest.test_offset_expiration
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d69c2583-32dd-4b13-b69c-14b4557123fc-->
<!--TEST RESULTS-->
<!--11d4d112-ced2-4592-b882-efb96940515d-->
<!--TEST RESULTS-->
<!--9f2561c3-30ac-46dc-b4b2-7c4eeaa1266c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73388">build#73388</a></summary>


| test_class          | test_method            | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history                                                                                                                                             |
|:--------------------|:-----------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------|
| OffsetRetentionTest | test_offset_expiration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73388#0199a1ba-fd52-4dc1-a77d-072a613f5795 | FAIL          | 0/1      |          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetRetentionTest&test_method=test_offset_expiration |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73435">build#73435</a></summary>


| test_class          | test_method            | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history                                                                                                                                             |
|:--------------------|:-----------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------|
| OffsetRetentionTest | test_offset_expiration | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73435#0199a41c-e361-4992-abc7-336cf1969927 | FAIL          | 0/1      |          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetRetentionTest&test_method=test_offset_expiration |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73469">build#73469</a></summary>


| test_class           | test_method          | test_arguments                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                            |
|:---------------------|:---------------------|:--------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi | test_no_auth_enabled | {"audit_transport_mode": "rpc"} | integration | https://buildkite.com/redpanda/redpanda/builds/73469#0199a5e7-f6c2-4e51-bbde-a0d0f162770d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled |
</details>
<!--/body-->

StephanDollberg commented:
/ci-repeat 1
tests/rptest/tests/offset_retention_test.py::OffsetRetentionTest.test_offset_expiration

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73435-->
<!--71a4e87a-38e4-4f00-8245-3fc7e5743f7f-->
<!--title-->
## Retry command for [Build#73435](https://buildkite.com/redpanda/redpanda/builds/73435)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_retention_test.py::OffsetRetentionTest.test_offset_expiration
```
<!--/body-->

StephanDollberg commented:
Failure is https://github.com/redpanda-data/redpanda/pull/27848

StephanDollberg commented:
pure rebase to get a green build
